File: cql3/selection/AbstractFunctionSelector.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.cassandra.cql3.selection;

import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.List;

import org.apache.commons.lang3.text.StrBuilder;

import org.apache.cassandra.cql3.functions.Function;
import org.apache.cassandra.db.marshal.AbstractType;
import org.apache.cassandra.exceptions.InvalidRequestException;

abstract class AbstractFunctionSelector<T extends Function> extends Selector
{
    protected final T fun;

    /**
     * The list used to pass the function arguments is recycled to avoid the cost of instantiating a new list
     * with each function call.
     */
    protected final List<ByteBuffer> args;
    protected final List<Selector> argSelectors;

    public static Factory newFactory(final Function fun, final SelectorFactories factories) throws InvalidRequestException
    {
        if (fun.isAggregate())
        {
            if (factories.doesAggregation())
                throw new InvalidRequestException("aggregate functions cannot be used as arguments of aggregate functions");
        }
        else
        {
            if (factories.doesAggregation() && !factories.containsOnlyAggregateFunctions())
                throw new InvalidRequestException(String.format("the %s function arguments must be either all aggregates or all none aggregates",
                                                                fun.name()));
        }

        return new Factory()
        {
            protected String getColumnName()
            {
                return new StrBuilder(fun.name().toString()).append('(')
                                                            .appendWithSeparators(factories.getColumnNames(), ", ")
                                                            .append(')')
                                                            .toString();
            }

            protected AbstractType<?> getReturnType()
            {
                return fun.returnType();
            }

            public boolean usesFunction(String ksName, String functionName)
            {
                return fun.usesFunction(ksName, functionName);
            }

            public Selector newInstance() throws InvalidRequestException
            {
                return fun.isAggregate() ? new AggregateFunctionSelector(fun, factories.newInstances())
                                         : new ScalarFunctionSelector(fun, factories.newInstances());
            }

            public boolean isWritetimeSelectorFactory()
            {
                return factories.containsWritetimeSelectorFactory();
            }

            public boolean isTTLSelectorFactory()
            {
                return factories.containsTTLSelectorFactory();
            }

            public boolean isAggregateSelectorFactory()
            {
                return fun.isAggregate() || factories.containsOnlyAggregateFunctions();
            }
        };
    }

    protected AbstractFunctionSelector(T fun, List<Selector> argSelectors)
    {
        this.fun = fun;
        this.argSelectors = argSelectors;
        this.args = Arrays.asList(new ByteBuffer[argSelectors.size()]);
    }

    public AbstractType<?> getType()
    {
        return fun.returnType();
    }

    @Override
    public String toString()
    {
        return new StrBuilder().append(fun.name())
                               .append("(")
                               .appendWithSeparators(argSelectors, ", ")
                               .append(")")
                               .toString();
    }
}

File: cql3/selection/abstract_function_selector.cc
/*
 * Modified by ScyllaDB
 * Copyright (C) 2015-present ScyllaDB
 */

/*
 * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
 */

#include "abstract_function_selector.hh"
#include "aggregate_function_selector.hh"
#include "scalar_function_selector.hh"
#include "to_string.hh"
#include "cql3/selection/selector_factories.hh"
#include "cql3/functions/abstract_function.hh"

namespace cql3 {

namespace selection {

bool abstract_function_selector::requires_thread() const {
    return _requires_thread;
}

shared_ptr<selector::factory>
abstract_function_selector::new_factory(shared_ptr<functions::function> fun, shared_ptr<selector_factories> factories) {
    if (fun->is_aggregate()) {
        if (factories->does_aggregation()) {
            throw exceptions::invalid_request_exception("aggregate functions cannot be used as arguments of aggregate functions");
        }
    } else {
        if (factories->does_aggregation() && !factories->contains_only_aggregate_functions()) {
            throw exceptions::invalid_request_exception(format("the {} function arguments must be either all aggregates or all none aggregates",
                                                            fun->name()));
        }
    }

    struct fun_selector_factory : public factory {
        shared_ptr<functions::function> _fun;
        shared_ptr<selector_factories> _factories;

        fun_selector_factory(shared_ptr<functions::function> fun,
                             shared_ptr<selector_factories> factories)
                : _fun(std::move(fun)), _factories(std::move(factories)) {
        }

        virtual sstring column_name() const override {
            return _fun->column_name(_factories->get_column_names());
        }

        virtual data_type get_return_type() const override {
            return _fun->return_type();
        }

        virtual shared_ptr<selector> new_instance() const override {
            using ret_type = shared_ptr<selector>;
            return _fun->is_aggregate() ? ret_type(::make_shared<aggregate_function_selector>(_fun, _factories->new_instances()))
                                        : ret_type(::make_shared<scalar_function_selector>(_fun, _factories->new_instances()));
        }

        virtual bool is_write_time_selector_factory() const override {
            return _factories->contains_write_time_selector_factory();
        }

        virtual bool is_ttl_selector_factory() const override {
            return _factories->contains_ttl_selector_factory();
        }

        virtual bool is_aggregate_selector_factory() const override {
            return _fun->is_aggregate() || _factories->contains_only_aggregate_functions();
        }

        virtual bool contains_only_simple_arguments() const override {
            return _factories->contains_only_simple_selection();
        }

        virtual bool is_count_selector_factory() const override {
            auto p = dynamic_cast<functions::abstract_function*>(_fun.get());
            if (!p) {
                return false;
            }
            return p->name().name == "countRows";
        }

        virtual bool is_reducible_selector_factory() const override {
            auto p = dynamic_cast<functions::aggregate_function*>(_fun.get());
            if (!p) {
                return false;
            }
            return p->is_reducible();
        }

        virtual std::optional<std::pair<query::forward_request::reduction_type, query::forward_request::aggregation_info>> 
        get_reduction() const override {
            auto p = dynamic_cast<functions::aggregate_function*>(_fun.get());
            if (!p) {
                return std::nullopt;
            }

            auto type = (p->name().name == "countRows") ? query::forward_request::reduction_type::count : query::forward_request::reduction_type::aggregate;
            auto info = query::forward_request::aggregation_info {
                .name = p->name(),
                .column_names = _factories->get_column_names()
            };

            return {{type, info}};
        }

    };

    return make_shared<fun_selector_factory>(std::move(fun), std::move(factories));
}

bool scalar_function_selector::requires_thread() const {
    return fun()->requires_thread();
}

}

}
